# Yelp Data Challenge

The Yelp Data Challenge can be found [Yelp Dataset Challenge](https://www.yelp.com/dataset_challenge).
<img src="./test/yelp_download.png">

Download JSON. 3.85GB of `yelp_dataset.rar`, and unrar to get 8.65GB of `yelp_dataset` folder.

In the folder, there are six `.json` files:

    business.json - 138MB
    checkin.json - 408MB
    photo.json - 25.7MB
    review.json - 5.35GB
    tip.json - 244.5MB
    user.json - 2.49GB

Note that each file is composed of a single object type, one json-object per-line. We used `business` and `review` dataset. We can take a look at some examples on [GitHub](https://github.com/Yelp/dataset-examples) to get us started. Here is a json record in business file and review file.

**business.json**
```Json
{
  "business_id": "1SWheh84yJXfytovILXOAQ",
  "name": "Arizona Biltmore Golf Club",
  "address": "2818 E Camino Acequia Drive",
  "city": "Phoenix",
  "state": "AZ",
  "postal_code": "85016",
  "latitude": 33.5221425,
  "longitude": -112.0184807,
  "stars": 3.0,
  "review_count": 5,
  "is_open": 0,
  "attributes": {
    "GoodForKids": "False"
  },
  "categories": "Golf, Active Life",
  "hours": null
}
```

**review.json**
```Json
{
  "review_id": "Q1sbwvVQXV2734tPgoKj4Q",
  "user_id": "hG7b0MtEbXx5QzbzE6C_VA",
  "business_id": "ujmEBvifdJM6h6RLv4wQIg",
  "stars": 1.0,
  "useful": 6,
  "funny": 1,
  "cool": 0,
  "text": "Total bill for this horrible service? Over $8Gs. These crooks actually had the nerve to charge us $69 for 3 pills. I checked online the pills can be had for 19 cents EACH! Avoid Hospital ERs at all costs.",
  "date": "2013-05-07 04:34:36"
}
```


# Process Business Data

First, process business dataset. Use `json.load` to read json data format. Then read it into `pandas.DataFrame`. We filtered the instances with **city contains Las Vegas** and **categories contains Restaurants**. At last, we only keep useful columns, such as **business_id**, **name**, **categories** and **stars**.

## Read Dataset

In [1]:
import json
import pandas as pd

# read dataset into DataFrame
with open('E:/GitHub/DS701-1904/data/yelp_dataset/business.json', encoding = 'utf-8') as f:
    df_business = pd.DataFrame(json.loads(line) for line in f)

In [2]:
df_business.head()

address  \
0     2818 E Camino Acequia Drive   
1            30 Eglinton Avenue W   
2       10110 Johnston Rd, Ste 15   
3   15655 W Roosevelt St, Ste 237   
4  4209 Stuart Andrew Blvd, Ste F   

                                          attributes             business_id  \
0                           {'GoodForKids': 'False'}  1SWheh84yJXfytovILXOAQ   
1  {'RestaurantsReservations': 'True', 'GoodForMe...  QXAEGFB4oINsVuTFxEYKFQ   
2  {'GoodForKids': 'True', 'NoiseLevel': 'u'avera...  gnKjwL_1w79qoiV3IC_xQQ   
3                                               None  xvX2CttrVhyG2z1dFg_0xw   
4  {'BusinessAcceptsBitcoin': 'False', 'ByAppoint...  HhyxOkGAM07SRYtlQ4wMFQ   

                                          categories         city  \
0                                  Golf, Active Life      Phoenix   
1  Specialty Food, Restaurants, Dim Sum, Imported...  Mississauga   
2                  Sushi Bars, Restaurants, Japanese    Charlotte   
3                      Insurance, Financial Services     Goodyear   
4  Plumbing, Shopping, Local Services, Home Servi...    Charlotte   

                                               hours  is_open   latitude  \
0                                               None        0  33.522143   
1  {'Monday': '9:0-0:0', 'Tuesday': '9:0-0:0', 'W...        1  43.605499   
2  {'Monday': '17:30-21:30', 'Wednesday': '17:30-...        1  35.092564   
3  {'Monday': '8:0-17:0', 'Tuesday': '8:0-17:0', ...        1  33.455613   
4  {'Monday': '7:0-23:0', 'Tuesday': '7:0-23:0', ...        1  35.190012   

    longitude                             name postal_code  review_count  \
0 -112.018481       Arizona Biltmore Golf Club       85016             5   
1  -79.652289       Emerald Chinese Restaurant     L5R 3E7           128   
2  -80.859132      Musashi Japanese Restaurant       28210           170   
3 -112.395596  Farmers Insurance - Paul Lorenz       85338             3   
4  -80.887223              Queen City Plumbing       28217             4   

   stars state  
0    3.0    AZ  
1    2.5    ON  
2    4.0    NC  
3    5.0    AZ  
4    4.0    NC

In [3]:
df_business.shape

(192609, 14)

## Filter by city and categories

We get 192609 business entities in dataset. Some of them are restaurants, some are not. We focus on the Las Vegas restaurants in Yelp. Thus, creat some filters, of which the business 
* is located in "Las Vegas"
* has no null category value
* contains "Restaurants" in their category

In [33]:
# create Pandas DataFrame filters

# city, check if it contains "Las Vegas"
filter_city = df_business['city'].apply(str).str.contains("Las Vegas")

# categories, check if it is null
filter_category_not_null = ~df_business["categories"].isnull()

# categories, check if it contains "Restaurants"
filter_category_restaurant = df_business["categories"].apply(str).str.contains("Restaurants")

# filter DataFrame, and name it df_filtered
df_filtered = df_business[filter_city & filter_category_not_null & filter_category_restaurant]

In [36]:
df_filtered.head()

address  \
17       1775 E Tropicana Ave, Ste 29   
25              6055 E Lake Mead Blvd   
75            6125 Spring Mountain Rd   
111                  4343 N Rancho Dr   
135  Artisan Hotel, 1501 W Sahara Ave   

                                            attributes  \
17   {'OutdoorSeating': 'False', 'BusinessAcceptsCr...   
25   {'BikeParking': 'True', 'BusinessParking': '{'...   
75   {'RestaurantsPriceRange2': '1', 'Ambience': '{...   
111                                               None   
135  {'RestaurantsAttire': ''dressy'', 'Corkage': '...   

                business_id  \
17   PZ-LZzSlhSe9utkQYU8pFg   
25   tstimHoMcYbkSC4eBA1wEg   
75   kANF0dbeoW34s2vwh6Umfw   
111  X8mtoSxY8whtmbDlj0D3Aw   
135  bJP4l_BGq2CudEu0m-wNjg   

                                            categories       city  \
17                                Restaurants, Italian  Las Vegas   
25   Mexican, Restaurants, Patisserie/Cake Shop, Fo...  Las Vegas   
75   Fast Food, Food, Restaurants, Ice Cream & Froz...  Las Vegas   
111                               Restaurants, Chinese  Las Vegas   
135        Restaurants, Pizza, Italian, American (New)  Las Vegas   

                                                 hours  is_open   latitude  \
17                                                None        0  36.100016   
25   {'Monday': '11:0-21:0', 'Tuesday': '10:0-21:0'...        1  36.195615   
75                                                None        0  36.125031   
111                                               None        1  36.238596   
135  {'Monday': '16:0-0:0', 'Tuesday': '16:0-0:0', ...        0  36.143672   

      longitude                                 name postal_code  \
17  -115.128529           Carluccio's Tivoli Gardens       89119   
25  -115.040529  Maria's Mexican Restaurant & Bakery       89156   
75  -115.225620                          Dairy Queen       89146   
111 -115.233331                Imperial Asian Buffet       89030   
135 -115.169792             Artisan Fine Dining Room       89102   

     review_count  stars state  
17             40    4.0    NV  
25            184    4.5    NV  
75             33    2.0    NV  
111             4    2.0    NV  
135             3    2.0    NV

In [37]:
df_filtered.shape # now we get 6812 records 

(6812, 14)

## Keep relevant columns

There are only a few columns that are useful.
* business_id
* name
* categories
* stars

We make a DataFrame that contains only the above mentioned columns, and name it as `df_selected_business`. In order to avoid naming conflicts with review dataset, we rename the column "stars" to "avg_stars".

In [41]:
# filter
df_selected_business = df_filtered[['business_id', 'name', 'categories', 'stars']]

# Rename
df_selected_business.rename(columns={"stars":"avg_stars"}, inplace=True)

In [42]:
df_selected_business.head()

business_id                                 name  \
17   PZ-LZzSlhSe9utkQYU8pFg           Carluccio's Tivoli Gardens   
25   tstimHoMcYbkSC4eBA1wEg  Maria's Mexican Restaurant & Bakery   
75   kANF0dbeoW34s2vwh6Umfw                          Dairy Queen   
111  X8mtoSxY8whtmbDlj0D3Aw                Imperial Asian Buffet   
135  bJP4l_BGq2CudEu0m-wNjg             Artisan Fine Dining Room   

                                            categories  avg_stars  
17                                Restaurants, Italian        4.0  
25   Mexican, Restaurants, Patisserie/Cake Shop, Fo...        4.5  
75   Fast Food, Food, Restaurants, Ice Cream & Froz...        2.0  
111                               Restaurants, Chinese        2.0  
135        Restaurants, Pizza, Italian, American (New)        2.0

## Save Results

Save the results to csv file, in case we need to reuse it in other notebooks.

In [46]:
# Save to your data folder for the next task
df_selected_business.to_csv("E:/GitHub/DS701-1904/data/selected_business.csv", index=False, encoding="utf-8")

In [19]:
# # Try reload the csv file to check if everything works fine
# pd.read_csv("../data/selected_business.csv", encoding="utf-8").head()

# Process Review Data

We make a DataFrame that contains and only contains the reviews about the business entities we just obtained.

**WARNING:** It is 5 GB review dataset. 

## Read Dataset

In [49]:
# Warning!!! 5GB review data
#with open('E:/GitHub/DS701-1904/data/yelp_dataset/review.json', encoding='utf-8') as f:
    #df_review = pd.DataFrame(json.loads(line) for line in f)

In [50]:
df_review.shape

(6685900, 9)

In [51]:
df_review_copy = df_review.copy()

## Join on business_id

We only keep reviews from last 1 years (it is better to perform filter before join operation),
* Otherwise your laptop may crush on memory when running machine learning algorithms
* Purposefully ignoring the reviews made too long time ago

To implement this join operation, we should 
* Prepare the business dataframe and set index to column "business_id", and name it as df_left
* Prepare the review dataframe and set index to column "business_id", and name it as df_right
* Also filter date after '2018-02-01'
* Inner join left and right table

In [52]:
# left table
df_left = df_selected_business.set_index('business_id')
# right table
df_review = df_review[df_review['date']>'2018-02-01'].set_index('business_id')
# inner join df_left and df_right
df_final = df_left.join(df_review,how='inner')
# reset the index 
df_final = df_final.reset_index()

In [53]:
df_final.shape

(193487, 12)

## Save Final Results

In [55]:
# Save to data/last_1_years_restaurant_reviews.csv for the next task
df_final.to_csv('E:/GitHub/DS701-1904/data/last_1_years_restaurant_reviews.csv',index=False)